File: backend/app/schemas/gl_catalog.py
"""
Pydantic shapes for the GL Codes (chart of accounts) API.

Mirrors `invoice_template.py` in shape so the frontend gets a
consistent rhythm across the two builders (templates / catalogs).

Response shapes:
  * `GLCatalogEntry`   — one GL row inside a catalog's `entries` array.
  * `GLCatalogOut`     — full record (returned by GET / POST / PATCH).
  * `GLCatalogSummary` — list-row payload (entry_count instead of the
                          full entries array — keeps the rail's GET
                          cheap when the user has many catalogs).
  * `GLCatalogDefault` — canonical built-in catalog the frontend uses
                          as a starter draft when nothing is saved yet.
                          Not persisted.
  * `ParsedGLUpload`   — output of `POST /gl-catalogs/parse-upload`. A
                          one-shot parse: returns the RAW source columns
                          and rows from the uploaded file plus a
                          `suggested_mapping` (canonical-field → source-
                          column hint). The frontend then runs an
                          explicit column-mapping step before assembling
                          the canonical entries — the parser deliberately
                          does NOT pre-build entries, since BillsIQ's
                          canonical schema is the source of truth and
                          the user must confirm (or override) the
                          mapping. The file itself isn't stored.
  * `GLUploadMapping`  — canonical-field → source-column-name dict the
                          mapping step works against. Sentinel `None`
                          means "no source column maps to this field".

`source` is a free-text Literal — adding a new origin later needs a
one-line schema change but no DB migration (the model uses a
free-text column).
"""

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, field_validator

# Where a catalog originated. Informational only.
GLCatalogSourceLiteral = Literal["default", "blank", "from_upload", "custom"]

# Sensible bounds. 1 entry minimum (a 0-row catalog is a placeholder,
# not a chart of accounts). 5000 max keeps the JSONB payload bounded
# while leaving headroom for any realistic chart we'll see — the
# canonical multifamily charts top out around a few hundred codes.
MIN_ENTRIES = 1
MAX_ENTRIES = 5000

MAX_CODE_LENGTH = 64
MAX_DESCRIPTION_LENGTH = 500
MAX_CATEGORY_LENGTH = 120
MAX_NOTES_LENGTH = 1000

# Hard cap on data rows the parser surfaces in `ParsedGLUpload.source_rows`.
# Mirrors MAX_ENTRIES so a parsed upload never contains more rows than a
# saved catalog can hold — prevents the "upload 50k-row file, mapping
# step looks fine, save fails at the entries-length validator" cliff.
MAX_PARSE_ROWS = MAX_ENTRIES


class GLCatalogEntry(BaseModel):
    """One GL row inside a catalog's `entries` array."""

    # Stable client-side id so the React editor can use it as a key the
    # moment it adds a row, before any server round-trip. The backend
    # only enforces uniqueness within the catalog (not strict UUID
    # format) so `e-0` style ids from a fresh draft are accepted.
    id: str = Field(min_length=1, max_length=64)

    # The GL account code (e.g. "5100", "6010-01", "1100.10"). Stored
    # as a free-text string — different orgs use different conventions
    # (numeric, dotted, hierarchical) and we don't want the model to
    # privilege any one of them.
    code: str = Field(min_length=1, max_length=MAX_CODE_LENGTH)

    # Human-readable description of the account. Required (a code with
    # no description is borderline useless — and uploaded charts always
    # have one).
    description: str = Field(min_length=1, max_length=MAX_DESCRIPTION_LENGTH)

    # Optional category / group label, e.g. "Operating Expenses",
    # "Asset", "Income". Free text — different ResMan exports use
    # slightly different vocabulary; we don't validate.
    category: str | None = Field(default=None, max_length=MAX_CATEGORY_LENGTH)

    # Whether this account should currently be considered usable.
    # Defaults to True so a new entry is "live" by default; users can
    # mark legacy / closed accounts inactive without deleting them.
    active: bool = True

    # Optional internal notes — useful for explaining unusual codes
    # to teammates ("use only for HOA pass-throughs", etc).
    notes: str | None = Field(default=None, max_length=MAX_NOTES_LENGTH)

    @field_validator("code")
    @classmethod
    def _strip_code(cls, v: str) -> str:
        v = v.strip()
        if not v:
            raise ValueError("GL code cannot be blank")
        return v

    @field_validator("description")
    @classmethod
    def _strip_description(cls, v: str) -> str:
        v = v.strip()
        if not v:
            raise ValueError("GL description cannot be blank")
        return v

    @field_validator("category")
    @classmethod
    def _strip_category(cls, v: str | None) -> str | None:
        if v is None:
            return None
        v = v.strip()
        return v or None

    @field_validator("notes")
    @classmethod
    def _strip_notes(cls, v: str | None) -> str | None:
        if v is None:
            return None
        v = v.strip()
        return v or None


def _validate_entry_invariants(
    entries: list[GLCatalogEntry],
) -> list[GLCatalogEntry]:
    """Shared invariants for create + update + parse-upload responses.

    * Entry ids must be unique within the catalog.
    * GL codes must be unique within the catalog (case-insensitive) —
      same code twice is almost always a data-entry mistake and we'd
      rather catch it at save time than during downstream validation.
    """
    ids = [e.id for e in entries]
    if len(ids) != len(set(ids)):
        raise ValueError("entry ids must be unique within the catalog")
    codes_lower = [e.code.strip().lower() for e in entries]
    if len(codes_lower) != len(set(codes_lower)):
        raise ValueError("GL codes must be unique within the catalog")
    return entries


class GLCatalogCreate(BaseModel):
    """Body for POST /gl-catalogs."""

    name: str = Field(min_length=1, max_length=255)
    description: str | None = None
    entries: list[GLCatalogEntry] = Field(
        min_length=MIN_ENTRIES, max_length=MAX_ENTRIES
    )
    source: GLCatalogSourceLiteral = "custom"

    @field_validator("name")
    @classmethod
    def _strip_name(cls, v: str) -> str:
        v = v.strip()
        if not v:
            raise ValueError("name cannot be blank")
        return v

    @field_validator("entries")
    @classmethod
    def _check_entries(cls, v: list[GLCatalogEntry]) -> list[GLCatalogEntry]:
        return _validate_entry_invariants(v)


class GLCatalogUpdate(BaseModel):
    """
    Body for PATCH /gl-catalogs/{id}. Every field optional.

    Sending `entries` REPLACES the array (no per-row patching). The
    frontend always sends the full new ordered list — that matches how
    a structural editor naturally batches edits and avoids the
    complexity of diff-based merges. Same contract as InvoiceTemplate.
    """

    name: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = None
    entries: list[GLCatalogEntry] | None = Field(
        default=None, min_length=MIN_ENTRIES, max_length=MAX_ENTRIES
    )
    # `source` is set at create time and not editable via PATCH.

    @field_validator("name")
    @classmethod
    def _strip_name(cls, v: str | None) -> str | None:
        if v is None:
            return None
        v = v.strip()
        if not v:
            raise ValueError("name cannot be blank")
        return v

    @field_validator("entries")
    @classmethod
    def _check_entries(
        cls, v: list[GLCatalogEntry] | None
    ) -> list[GLCatalogEntry] | None:
        if v is None:
            return None
        return _validate_entry_invariants(v)


class GLCatalogOut(BaseModel):
    id: uuid.UUID
    name: str
    description: str | None
    entries: list[GLCatalogEntry]
    source: GLCatalogSourceLiteral
    created_by: uuid.UUID | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class GLCatalogSummary(BaseModel):
    """List-row shape — no entries array, just its size."""

    id: uuid.UUID
    name: str
    description: str | None
    source: GLCatalogSourceLiteral
    entry_count: int
    created_at: datetime
    updated_at: datetime


class GLCatalogList(BaseModel):
    items: list[GLCatalogSummary] = Field(default_factory=list)


class GLCatalogDefault(BaseModel):
    """
    Canonical built-in catalog the frontend uses as a starting editable
    draft when nothing is saved yet. Not persisted.
    """

    name: str
    description: str | None
    entries: list[GLCatalogEntry]


class GLUploadMapping(BaseModel):
    """
    Mapping from BillsIQ's canonical GL fields to source column names
    in the uploaded file.

    Each value is either:
      * the EXACT NAME of a column in `ParsedGLUpload.source_columns`
        (the frontend uses these strings as the value attribute on
        each canonical field's <select>), or
      * `None`, meaning "no source column maps to this canonical field".

    Required-field omissions (`code` / `description` left as None) are
    enforced by the frontend at mapping-confirm time and again by the
    `GLCatalogCreate` validator when the catalog is finally saved —
    this schema itself doesn't enforce them, since a partially-mapped
    state is valid mid-wizard.

    Backend uses this shape only for the `suggested_mapping` field of
    `ParsedGLUpload` (a hint the frontend may accept or override). The
    final user-confirmed mapping never round-trips through the API —
    the frontend applies it locally to assemble canonical entries that
    are then sent to `POST /gl-catalogs`.
    """

    code: str | None = None
    description: str | None = None
    category: str | None = None
    active: str | None = None
    notes: str | None = None


class ParsedGLUpload(BaseModel):
    """
    Result of `POST /gl-catalogs/parse-upload`.

    One-shot parse: the file isn't stored. The parser surfaces the
    raw source data (header columns + data rows) and a best-effort
    `suggested_mapping` from BillsIQ's canonical GL fields to source
    column names. The frontend then runs an EXPLICIT column-mapping
    step where the user confirms or overrides the suggestion before
    the canonical entries are assembled and sent to POST /gl-catalogs.

    Why we deliberately don't pre-build entries here:
      * BillsIQ's canonical GL schema (code/description/category/
        active/notes) is the source of truth. A raw uploaded file's
        column names can be anything — the user must explicitly
        confirm the mapping before any row becomes a canonical entry.
      * Parsing-as-mapping silently couples header detection to entry
        creation and hides misdetections from the user. Splitting the
        two steps surfaces the mapping decision as a first-class one.

    Soft failures (no header row, unsupported file shape) come back
    as a successful response with empty `source_columns` /
    `source_rows` and a `parse_warning` explaining what happened —
    the modal renders a hint and lets the user pick a different file.
    """

    filename: str
    detected_format: Literal["csv", "xlsx"]
    source_columns: list[str]
    """Column headers as they appeared in the uploaded file, in order.
    Empty list when no header row was recognized — `parse_warning`
    explains why and the frontend's mapping step is short-circuited."""
    source_rows: list[list[str]]
    """Data rows below the header, each padded / truncated to
    `len(source_columns)` so column index N is meaningful for every
    row. Capped at `MAX_PARSE_ROWS`."""
    suggested_mapping: GLUploadMapping
    """Best-effort canonical-field → source-column-name guess. Always
    set (sentinel `None` for fields the parser couldn't match), so
    the frontend can pre-populate dropdowns the user then confirms
    or overrides. Never silently committed — the user always sees
    and can change every choice."""
    parse_warning: str | None = None
    """Set when the parser still returned a response but something is
    worth telling the user about — e.g. no recognizable header row,
    or the file produced more rows than `MAX_PARSE_ROWS` and was
    truncated."""


# ---------------------------------------------------------------------------
# Default catalog — a small, sensible starter
# ---------------------------------------------------------------------------
#
# This is the canonical chart the frontend renders as a draft on first
# landing. Sized to be useful (not just a single-row toy) but small
# enough that the user can scan it and decide what to keep / edit.
#
# Codes follow the common multifamily property-management 4-digit
# convention: 1xxx = assets, 2xxx = liabilities, 3xxx = equity,
# 4xxx = income, 5xxx-6xxx = operating expenses. Adjust cautiously —
# every fresh draft starts from this list, so adding/removing entries
# affects the "first GL catalog" experience.

_DEFAULT_ENTRIES: tuple[tuple[str, str, str], ...] = (
    # (code, description, category)
    ("1100", "Cash - Operating", "Asset"),
    ("1200", "Accounts Receivable", "Asset"),
    ("2100", "Accounts Payable", "Liability"),
    ("2200", "Accrued Expenses", "Liability"),
    ("4000", "Rental Income", "Income"),
    ("4100", "Other Income", "Income"),
    ("5100", "Repairs & Maintenance", "Operating Expense"),
    ("5110", "HVAC Maintenance", "Operating Expense"),
    ("5120", "Plumbing", "Operating Expense"),
    ("5130", "Electrical", "Operating Expense"),
    ("5200", "Cleaning & Janitorial", "Operating Expense"),
    ("5300", "Landscaping", "Operating Expense"),
    ("5400", "Pest Control", "Operating Expense"),
    ("5500", "Utilities - Common Area", "Operating Expense"),
    ("5600", "Property Insurance", "Operating Expense"),
    ("5700", "Property Tax", "Operating Expense"),
    ("6100", "Office Supplies", "Operating Expense"),
    ("6200", "Marketing & Advertising", "Operating Expense"),
    ("6300", "Professional Fees", "Operating Expense"),
    ("6400", "Bank Fees", "Operating Expense"),
)


def build_default_catalog() -> GLCatalogDefault:
    """Construct the canonical default catalog. Pure function — no I/O."""
    entries = [
        GLCatalogEntry(
            id=f"d-{i}",
            code=code,
            description=desc,
            category=category,
            active=True,
            notes=None,
        )
        for i, (code, desc, category) in enumerate(_DEFAULT_ENTRIES)
    ]
    return GLCatalogDefault(
        name="Default GL Catalog",
        description=(
            "Built-in starter chart of accounts covering common "
            "multifamily property accounting categories. Edit freely — "
            "saving creates an independent copy you can keep refining."
        ),
        entries=entries,
    )

File: frontend/src/types/gl-catalog.ts
/**
 * GL Codes (chart of accounts) — types mirror `app/schemas/gl_catalog.py`.
 *
 * One GLCatalog is a saved, named chart of accounts the user maintains
 * for ResMan invoice processing. Decoupled from `InvoiceTemplate`:
 *
 *   * InvoiceTemplate owns the COLUMN SHAPE of an invoice export.
 *   * GLCatalog owns the ROWS of a chart of accounts (one per GL code)
 *     the user later validates / maps invoice line-items against.
 *
 * Same builder pattern across both: list rail + center editor +
 * creation modal. Same source-of-origin metadata. Same JSONB-backed
 * editable list on the server.
 */

/**
 * Where this catalog was originally seeded. Informational — drives a
 * small "origin" hint in the UI but has no functional effect.
 */
export type GLCatalogSource =
  | "default"
  | "blank"
  | "from_upload"
  | "custom";

/**
 * One entry inside a catalog's `entries` array.
 *
 * `id` is a stable client-side string key (not a numeric index) so
 * reordering / mid-edit React reconciliation stays correct. Generated
 * client-side when adding a row; the backend only enforces uniqueness
 * within the catalog.
 *
 * `code` is free text (`"5100"`, `"6010-01"`, `"1100.10"`) — different
 * orgs use different conventions and we don't want to privilege one.
 *
 * `category` is a free-text group label ("Operating Expense",
 * "Asset", "Income"). Optional.
 *
 * `active` defaults to true; users can flip legacy/closed accounts to
 * inactive without deleting them.
 *
 * `notes` is for internal context — "use only for HOA pass-throughs",
 * etc.
 */
export interface GLCatalogEntry {
  id: string;
  code: string;
  description: string;
  category: string | null;
  active: boolean;
  notes: string | null;
}

export interface GLCatalogOut {
  id: string;
  name: string;
  description: string | null;
  entries: GLCatalogEntry[];
  source: GLCatalogSource;
  created_by: string | null;
  created_at: string;
  updated_at: string;
}

export interface GLCatalogSummary {
  id: string;
  name: string;
  description: string | null;
  source: GLCatalogSource;
  entry_count: number;
  created_at: string;
  updated_at: string;
}

export interface GLCatalogList {
  items: GLCatalogSummary[];
}

/**
 * Built-in canonical catalog returned by GET /defaults/canonical. Not
 * persisted — the frontend uses it as an editable starter draft.
 */
export interface GLCatalogDefault {
  name: string;
  description: string | null;
  entries: GLCatalogEntry[];
}

/** POST body — all top-level fields required. */
export interface GLCatalogCreate {
  name: string;
  description?: string | null;
  entries: GLCatalogEntry[];
  source?: GLCatalogSource;
}

/**
 * PATCH body. Every field optional; omitting a field leaves the
 * persisted value unchanged. Sending `entries` REPLACES the array
 * outright — the editor always sends the full new ordered list.
 * Sending `null` for `description` clears it.
 */
export interface GLCatalogUpdate {
  name?: string;
  description?: string | null;
  entries?: GLCatalogEntry[];
}

/**
 * Mapping from BillsIQ's canonical GL fields to source column names
 * in an uploaded chart-of-accounts file.
 *
 * Each value is either:
 *   * the EXACT NAME of a column in `ParsedGLUpload.source_columns`
 *     (matching the string the user picks in the field's dropdown), or
 *   * `null`, meaning "no source column maps to this canonical field".
 *
 * Mid-mapping, leaving a required field as `null` is fine — the
 * mapping step's "Create catalog" button is gated on every required
 * field being mapped before the canonical entries are assembled.
 *
 * Mirrors `app.schemas.gl_catalog.GLUploadMapping`.
 */
export interface GLUploadMapping {
  code: string | null;
  description: string | null;
  category: string | null;
  active: string | null;
  notes: string | null;
}

/**
 * Result of POST /gl-catalogs/parse-upload — a one-shot parse of a
 * csv/xlsx GL chart upload.
 *
 * The parser surfaces RAW source data (header columns + data rows) and
 * a best-effort `suggested_mapping` from the canonical BillsIQ GL
 * fields to source column names. The frontend then runs an explicit
 * column-mapping step where the user confirms or overrides each
 * suggestion before the canonical entries are assembled.
 *
 * The parser deliberately does NOT pre-build entries — BillsIQ's
 * canonical schema is the source of truth and the user must explicitly
 * map source columns to canonical fields. The file itself is NOT
 * stored server-side.
 */
export interface ParsedGLUpload {
  filename: string;
  detected_format: "csv" | "xlsx";
  /** Header cells from the uploaded file, in order. Empty when no
   * recognizable header row was found. */
  source_columns: string[];
  /** Data rows below the header, each padded / truncated to
   * `source_columns.length` so column index N is meaningful for every
   * row. Capped server-side at MAX_ENTRIES. */
  source_rows: string[][];
  /** Best-effort canonical-field → source-column-name guess. Always
   * shown to the user on the mapping step — never silently committed.
   * Sentinel `null` means the parser couldn't match that field. */
  suggested_mapping: GLUploadMapping;
  /** Set when the parser still returned a response but something is
   * worth telling the user about — e.g. no recognizable header could
   * be identified, or rows were truncated to MAX_ENTRIES. */
  parse_warning: string | null;
}

/** Bounds enforced by the backend Pydantic schema. */
export const MIN_ENTRIES = 1;
export const MAX_ENTRIES = 5000;
export const MAX_CODE_LENGTH = 64;
export const MAX_DESCRIPTION_LENGTH = 500;
export const MAX_CATEGORY_LENGTH = 120;
export const MAX_NOTES_LENGTH = 1000;

/**
 * Generate a stable client-side id for a freshly-added entry. Uses
 * `crypto.randomUUID` where available, falls back to a Math.random
 * hex so the entry-id contract holds in environments without a crypto
 * API. The backend validates uniqueness within the catalog, not the
 * format — `e-0` style ids from a fresh draft are accepted.
 */
export function newEntryId(): string {
  if (typeof crypto !== "undefined" && "randomUUID" in crypto) {
    return crypto.randomUUID();
  }
  return `e-${Math.random().toString(16).slice(2, 10)}-${Date.now().toString(
    16,
  )}`;
}

// ---------------------------------------------------------------------------
// Canonical GL fields — single source of truth for the mapping UI
// ---------------------------------------------------------------------------

/**
 * Keys of the canonical GL schema, in the order the mapping UI lists
 * them. Matches `GLUploadMapping`'s field names.
 */
export type CanonicalGLFieldKey =
  | "code"
  | "description"
  | "category"
  | "active"
  | "notes";

/**
 * Describes one canonical BillsIQ GL field for the mapping step.
 *
 * `required: true` means the mapping step's "Create catalog" button
 * stays disabled until the user picks a source column for this field.
 * Optional fields may be left unmapped; the canonical entry's value
 * for that field becomes the documented default (null for nullable
 * strings, true for `active`).
 */
export interface CanonicalGLField {
  key: CanonicalGLFieldKey;
  label: string;
  required: boolean;
  /** One-sentence hint shown under the dropdown. */
  hint: string;
}

export const CANONICAL_GL_FIELDS: readonly CanonicalGLField[] = [
  {
    key: "code",
    label: "GL Code",
    required: true,
    hint: "The unique account number — e.g. 5100, 6010-01, 1100.10.",
  },
  {
    key: "description",
    label: "Description",
    required: true,
    hint: "Human-readable account name.",
  },
  {
    key: "category",
    label: "Category / Group",
    required: false,
    hint: "Optional grouping like 'Operating Expense', 'Asset', 'Income'.",
  },
  {
    key: "active",
    label: "Active / Inactive",
    required: false,
    hint:
      "Optional status flag. Recognizes yes/no, true/false, 1/0, " +
      "active/inactive. Missing values default to active.",
  },
  {
    key: "notes",
    label: "Notes",
    required: false,
    hint: "Optional internal context.",
  },
];

/** Empty mapping — every canonical field unmapped. */
export function emptyGLUploadMapping(): GLUploadMapping {
  return {
    code: null,
    description: null,
    category: null,
    active: null,
    notes: null,
  };
}

// ---------------------------------------------------------------------------
// Active-cell interpretation
// ---------------------------------------------------------------------------

const ACTIVE_TRUE_TOKENS = new Set([
  "yes",
  "y",
  "true",
  "t",
  "1",
  "active",
  "enabled",
  "open",
]);

const ACTIVE_FALSE_TOKENS = new Set([
  "no",
  "n",
  "false",
  "f",
  "0",
  "inactive",
  "disabled",
  "closed",
  "archived",
  "deleted",
]);

/**
 * Map a string from the optional `active` source column to a bool.
 *
 * Defaults to `true` when the cell is empty or unrecognized — a
 * missing status almost always means "active" in the source systems
 * we've seen, and the user can flip individual rows in the editor
 * after import.
 *
 * Lives here (rather than in the backend parser) because the column
 * the user maps to `active` isn't known until the mapping step
 * confirms it — the same source value ("Open") might be a status in
 * one upload and a category in another.
 */
export function interpretActive(raw: string | null | undefined): boolean {
  if (!raw) return true;
  const norm = raw.trim().toLowerCase();
  if (!norm) return true;
  if (ACTIVE_FALSE_TOKENS.has(norm)) return false;
  if (ACTIVE_TRUE_TOKENS.has(norm)) return true;
  return true;
}

// ---------------------------------------------------------------------------
// Mapping → canonical entries
// ---------------------------------------------------------------------------

/**
 * Result of applying a `GLUploadMapping` to the parser's source rows.
 *
 * `entries` is what gets sent to POST /gl-catalogs (after a final
 * de-dupe pass — see `dedupeAndCount` below). The counters surface
 * data-quality information the modal renders before the user
 * commits, so misdetections (mapping a "type" column to `code` and
 * getting 200 duplicates) are visible up front.
 */
export interface MappedSourceRows {
  entries: GLCatalogEntry[];
  blankRequiredCount: number;
  duplicateCodeCount: number;
}

/**
 * Apply a confirmed mapping to the parser's source rows, producing
 * canonical `GLCatalogEntry`s.
 *
 * Behaviour:
 *   * Rows whose mapped `code` is blank are dropped and counted as
 *     `blankRequiredCount` (`code` is required by the canonical
 *     schema; a row without it can't become a valid entry).
 *   * Rows whose mapped `description` is blank fall back to the code
 *     so the entry is at least constructible — the user can fix it in
 *     the editor after import.
 *   * Codes are de-duplicated case-insensitively; first occurrence
 *     wins and subsequent duplicates are counted as
 *     `duplicateCodeCount`. Matches the backend's uniqueness rule on
 *     `GLCatalogCreate`.
 *   * `active` cells are interpreted via `interpretActive`. When no
 *     source column is mapped, every entry gets `active: true`.
 *   * `category` and `notes` collapse blank → `null` to match the
 *     canonical schema's nullable string convention.
 *
 * `code` and `description` are the only required canonical fields;
 * the function still runs when other mappings are `null` — it just
 * leaves the corresponding canonical field at its default.
 */
export function applyMappingToSourceRows(
  sourceColumns: readonly string[],
  sourceRows: readonly (readonly string[])[],
  mapping: GLUploadMapping,
): MappedSourceRows {
  // Resolve column NAMES → INDICES once. Unknown names (or `null`)
  // map to -1, treated as "no source column for this field".
  const resolveIdx = (col: string | null): number => {
    if (!col) return -1;
    const idx = sourceColumns.indexOf(col);
    return idx >= 0 ? idx : -1;
  };

  const codeIdx = resolveIdx(mapping.code);
  const descIdx = resolveIdx(mapping.description);
  const catIdx = resolveIdx(mapping.category);
  const activeIdx = resolveIdx(mapping.active);
  const notesIdx = resolveIdx(mapping.notes);

  const cellAt = (row: readonly string[], idx: number): string => {
    if (idx < 0 || idx >= row.length) return "";
    return (row[idx] ?? "").trim();
  };

  const seenCodes = new Set<string>();
  const entries: GLCatalogEntry[] = [];
  let blankRequiredCount = 0;
  let duplicateCodeCount = 0;

  for (const row of sourceRows) {
    const code = cellAt(row, codeIdx);
    if (!code) {
      // No code → no canonical entry. Counted so the UI can warn.
      blankRequiredCount += 1;
      continue;
    }

    const codeKey = code.toLowerCase();
    if (seenCodes.has(codeKey)) {
      duplicateCodeCount += 1;
      continue;
    }
    seenCodes.add(codeKey);

    const description = cellAt(row, descIdx) || code;
    const categoryRaw = cellAt(row, catIdx);
    const notesRaw = cellAt(row, notesIdx);
    const activeRaw = activeIdx >= 0 ? cellAt(row, activeIdx) : null;

    entries.push({
      id: newEntryId(),
      code,
      description,
      category: categoryRaw || null,
      active: interpretActive(activeRaw),
      notes: notesRaw || null,
    });
  }

  return { entries, blankRequiredCount, duplicateCodeCount };
}
